### ISSUE TITLE
Joining large room over federation causes "weird state"

### ISSUE THREAD

**Issue by Valodim** [matrix-org/synapse#4803]
Joining a large-ish federated room (a label that applies to basically any bridged freenode room that's been around for a while) on my HS reproducibly leads to a rollercoaster of odd behaviors that are bewildering to users and troublesome for administration. However, the behavior does seem fairly consistent every time, I was able to observe it fully at least twice (in the last couple weeks).

Rough timeline of events:
* user joins a room over federation. concretely, this happened for at least #freenode_#esperanto:matrix.org and #freenode_#selenium:matrix.org.
* riot times out after a while, the room doesn't show up
* (at this point, the user possibly tries again to join the room? not sure if this happened every time, but it may or may not impact how synapse deals with the situation)
* after a short while, synapse enters a weird state: most things look normal, i.e. clients get sync answers in acceptable time, sending a message will work (grey text becomes black after a while), even notifications to messages are sent (and showing "X unread" bubbles as usual), everything looks mostly normal. **but**, messages do not actually get through. they just don't show up, neither on other devices of the sender, for recipients, or to appservices
* there are no suspicious log messages during this time, and RAM and CPU usage also does *not* spike super high.
* this weird state lasts for a long time. I don't know how long exactly or if it recovers up at some point, because I restarted synapse after ~an hour when this happened to avoid too much downtime. 
* after restarting synapse, the messages that weren't delivered before all show up, with correct timestamps as if they had been there all along. synapse still doesn't seem particularly stressed either.
* after many hours (10+, i.e. "not today"), the room shows up for the user who joined

I can't say if this happens only on my instance? As far as I'm aware, I'm not doing anything special on it.

It goes without saying though that this is very disruptive for users, probably worse than actual unresponsiveness because the behavior is so inexplicable.

**Comment by neilisfragile** [comment]
That does sound weird. Anything unusual in the logs?

**Comment by Valodim** [comment]
Nope, nothing :( Messages that users write *do* get through after a restart, so they are persisted. They are also marked as sent in Riot and cause notifications right away. So apparently they are processed almost entirely, they just never make it into syncs? No ideat

Luckily people don't join large rooms too often, otherwise this would be causing quite some headache for me hs's reliability

**Comment by Valodim** [comment]
Reporting in that this is still happening. Some user joined a large room (or rooms, not sure), symptoms as above. This is extremely unpleasant for UX, because everything still feels mostly normal and synapse responds normally to most client requests, but messages just do not actually get through.

**Comment by Valodim** [comment]
Reporting in: still happening, still extremely annoying.

**Comment by reivilibre** [comment]
I think the modern day lazy joins should make this a lot better. If you still have trouble, please open a new issue — sorry.
